design/ingress-status-loadbalancer.md

```diff
@@ -0,0 +1,101 @@
+# Ingress Status Loadbalancer Support
+
+Status: _Draft_
+
+This proposal describes how to add support for the Ingress.Status.LoadBalancer field to Contour.
+
+## Goals
+
+- Populate the status.loadbalancer field in all networking.k8s.io/v1beta.Ingress objects managed by Contour.
+- Contour pass the Ingress compliance tests for networking.k8s.io/v1beta.Ingress objects.
+
+## Non Goals
+
+- Similar functionality for IngressRoute (deprecated) and HTTPProxy are out of scope.
+
+## Background
+
+The Kubernetes Ingress object provides a field on the object's status to record address details for how access the ingresses virtualhosts from outside the cluster.
+This is known as the status loadbalancer information after the name of the fields on the Ingress Object.
+
+Details from the Kubernetes spec [are here](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#loadbalancerstatus-v1-core).
+
+## High-Level Design
+
+There are two components to this solution
+
+1. Discoverying the loadbalancer information
+2. Writing the loadbalacner information to the ingress objects.
+
+The former, the discovery will need to support 
+
+1. Hard coded IP/address details (actually a list, and it can be an address or an ip)
+2. Discoverying information from the service document.
+
+### Hard coded load balancer addresses
+
+While the hard coded approach sounds simpler, and is, it will likely not be the common use case, nor will it be the default.
+This is for several reasons.
+
+- The hard coded address model implies either the cluster is running on prem, or using host networking.
+- The hard coded IP model requires the administator to know the IP addresses of the _envoy_ service (not the contour service) which may change over time. 
+
+The configuration will likely go into Contour's configuration file.
+
+These restrictions (which are administrative, not ones we've placed on the solution) make this bookkeeping expensive, prohibitivly in the case of a highly dynamic cluster.
+
+### Dynamic load balancer address discovery
+
+The far more common case will be dynamic load balancer discovery.
+In this model Contour will track a service document (default to projectcontour/envoy, but configurable) and use those details as the list of ip addresses to write back to the ingress' status.
+
+The nginx controller goes to some lengths to try to dynamically figure out what it's status document as well as other side channels to discover the address if that fails.
+For this first implementation I propose we stick to the discovery methods outlined above.
+
```

Comment:
Is the discoverer not being able to discover some option a fatal error?

My feeling is no, I think, since that's the current situation anyway.

Reply:
> Is the discoverer not being able to discover some option a fatal error?

yeah, I'm thinking no as well. maybe we log something (as long as its not too incessant) but I can easily see cases where contour's pod races with the creation of envoy's service object and that'll just generate confusion and bug reports.

Reply:
There's also a bit of delay to determine the ip/name while the lb is spun up, so we need to make sure that delay is not a failure.

Reply:
As a compromise, how about we log when new address details become available, ie
```
ingressStatusWriter: loadbalancer address details changed to %v, updating Ingress objects
```

Reply:
Maybe a metric as well?